### ISSUE TITLE
openssl-3.0.13 issue with `rehash` after renaming cert file in trust store directory

### ISSUE THREAD

**Issue by sshedi** [openssl/openssl#23882]
System details:
```
Photon OS 5.0
openssl v3.0.13
```

Expected behaviour:
Self signed cert error should not show up after adding the pem file to /etc/ssl/certs and running rehash of openssl certs.

Actual behaviour:
Self signed cert error is showing up after adding pem file to /etc/ssl/certs but only if pem file name is changed.

Steps to reproduce:
Generate a self signed cert using:
```
openssl req -new -x509 -keyout cert.pem -out cert.pem -days 365 -nodes
```

Start a simple http server using:
```
#!/usr/bin/env python3

import http.server, ssl, socketserver
from http.server import SimpleHTTPRequestHandler, HTTPServer

addr = ("192.168.1.100", 4443)

httpd = HTTPServer(addr, SimpleHTTPRequestHandler)
httpd.socket = ssl.wrap_socket(httpd.socket,
                                   certfile="cert.pem", server_side=True)
httpd.serve_forever()
```

Copy `cert.pem` to `/etc/ssl/certs` and run rehash of certs
```
# rehash_ca_certificates.sh

root@ph5dev [ ~ ]# cat /bin/rehash_ca_certificates.sh
#!/bin/bash

set -e

CERTIFICATES_PATH="/etc/ssl/certs/"
CONCATENATED_CERT_FILE="/etc/pki/tls/certs/ca-bundle.crt"

if ! openssl rehash $CERTIFICATES_PATH; then
   echo "Error while c_rehashing"
fi

cat $CERTIFICATES_PATH*.pem > $CONCATENATED_CERT_FILE
```

Now open a new session on the same host and do:
```
openssl s_client -connect 192.168.1.100:4443 -showcerts
```

This time result will look something like:
```
root@ph5dev [ ~ ]# openssl s_client -connect 192.168.1.100:4443 -showcerts
CONNECTED(00000003)
Can't use SSL_get_servername
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify return:1
...
```

Now do `mv /etc/ssl/certs/cert.pem /etc/ssl/certs/x.pem` and run `rehash_ca_certificates.sh`again.

And if we do:
```
openssl s_client -connect 192.168.1.100:4443 -showcerts
```

We see:
```
root@ph5dev [ ~ ]# openssl s_client -connect 10.197.103.89:4443 -showcerts
CONNECTED(00000003)
Can't use SSL_get_servername
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify error:num=18:self-signed certificate ===========================> this is not expected
verify return:1
...
```

**Comment by sshedi** [comment]
This issue is not seen in `openssl-1.0.2zj`

**Comment by sshedi** [comment]
Can you please assist @t8m @DDvO

**Comment by halinwu** [comment]
I have the same problem , my vCenter vm need communicate with the Replication vm using the customize  CA , it couldn't work . photon os : 4

**Comment by RoLex** [comment]
+1 - same issue. Been trying 50 different ways to bypass using SSL_CTX_set_verify return 1, SSL_CTX_set_security_level(0), SSL_CTX_set_cert_verify_callback return 1, SSL_VERIFY_NONE, null callback, cipher ALL:@SECLEVEL=0, etc. Regardless of what server code does, library does not allow to bypass!

**Comment by DDvO** [comment]
@sshedi I just noticed your issue.

To reproduce, I started a test server on `localhost:4443` with Python 3.14.2 simply like this:
```
python -m http.server 4443 --tls-cert cert.pem --tls-key cert.pem
```

Then made `cert.pem` locally trusted like this:
```
mkdir certs
cp cert.pem certs/
openssl rehash certs/
```
Consequently, `certs/` contains the symlink
```
9da13359.0 -> cert.pem
```

This way, `s_client` succeeds:
```
echo | openssl s_client -connect localhost:4443 -verifyCApath certs/ -verify_return_error
```

Then essentially followed your steps:
```
mv  certs/{cert,x}.pem
openssl rehash certs/
```
Thus, `certs/` contains a further symlink
```
9da13359.1 -> x.pem
```
and `openssl s_client -connect localhost:4443 verifyCApath certs/ -verify_return_error` fails.

The reason being that `certs/` still contains the symlink 
```
9da13359.0 -> cert.pem
```
which is broken since `cert.pem` has been renamed.
In other words, the trust store directory is in an inconsistent state.

To fix this:
```
mv -f certs/9da13359.{1,0}
```

I wonder why you want(ed) to rename `cert.pem` there.

**Comment by sshedi** [comment]
Hi @DDvO you are right. Thank you.

In my latest script I added an instruction to remove dead symlinks.
https://github.com/vmware/photon/blob/5.0/SPECS/openssl/rehash_ca_certificates.sh#L13

And problem is not seen anymore. But I think this should happen automatically. Why consider a non existent file as valid.
I understand that this is a user error, just trying to understand the rationale behind this implementation.

**Comment by DDvO** [comment]
Yeah, since OpenSSL 3.0 (or already earlier), the code collecting trusted certs used by `s_client` is not robust/tolerant w.r.t. dangling symlinks. 
Good that you already implemented a solution/workaround for this.

I did not check with older (unsupported) OpenSSL versions, but this looks like a (minor) regression, likely by accident.

**Comment by sshedi** [comment]
Bu the weird part is, even after removing the dangling symlink the issue still happens.
I tested with `openssl-3.0.18`
openssl rehash removing the dangling symlink is a different part, let;s ignore it for now.

Cert.pem and x.pem have same data with different file names.
http server is loaded and running with cert.pem and x.pem is there in cert store but still "openssl s_client -connect <IP>:4443 -showcerts" fails. And it is happy only when both pem files have same file name, that is cert.pem

**Comment by sshedi** [comment]
Tested with `openssl-3.5.6` and the behavior is same.
Here is the python-3.14 compatible test script:
```
#!/usr/bin/env python3

import ssl
from http.server import SimpleHTTPRequestHandler, HTTPServer

addr = ("192.168.1.123", 4443)

httpd = HTTPServer(addr, SimpleHTTPRequestHandler)

context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
context.load_cert_chain(certfile="cert.pem")

httpd.socket = context.wrap_socket(
    httpd.socket,
    server_side=True
)

httpd.serve_forever()
```

**Comment by DDvO** [comment]
I meanwhile tried tracing relevant changes in `get_cert_by_subject_ex()` of `crypto/x509/by_dir.c`
and did not find anything that would explain why this worked earlier (as you reported) in OpenSSL 1.0.2.

Anyway, the root cause of the issue is that `stat()` fails on broken links and then the inner loop for a given subject hash value terminates, while there may be further symlinked candidate files with subsequent indexes.
The attached patch [graceful_get_cert_by_subject.diff.txt](https://github.com/user-attachments/files/27699821/graceful_get_cert_by_subject.diff.txt) makes the function behave gracefully on 
* broken symlinks (well, on non-Windows systems) and
* errors reading the respective cert/CRL from symlinked files.

**Comment by sshedi** [comment]
Thanks @DDvO the provided fix works, tested OK. Sorry for the delay, was busy in other tasks.
Hope this gets ported to 3.5.x LTS branch as well.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in fixing the bug or if you have reasons why this bug is important enough so we should consider implementing it with higher priority than other requests. The issue will be closed after 2 months otherwise.

**Comment by sshedi** [comment]
Hi @t8m, @DDvO provided a patch at https://github.com/siemens/openssl/pull/10, I tested it and it works fine. Are you looking for something better?

**Comment by t8m** [comment]
If @DDvO submits it as a regular PR, it can be considered.

**Comment by sshedi** [comment]
Hi @DDvO sorry for the spam, would you mind raising a PR for this?

**Comment by DDvO** [comment]
All right, provided the suggested patch (which as confirmed to work) as a PR in #31594.
